### PR TITLE
Gtk3

### PR DESCRIPTION
**Pull request by 01micko**
I'm attempting to use the correct options to `gtkdialog` to call a `css` file instead of my nasty hack of resetting `XDG_CONFIG_HOME`

Still a work in progress.

### CODE REVIEW

**Comment by peabee** [comment]
Sounds like a good idea as long as it doesn't break Gtk2 builds...... Not sure what we can do to help other than test builds once they appear .......??

**Comment by zigbert** [comment]
I am in lack of time to take a deep dive these days, but my brief look into the code looks very promising. I like the --styles solution which is clean and compatible with gtk2. There is still some issues to solve, but you have set the framework, which is the key for success.

**Comment by dimkr** [comment]
Triggered https://github.com/puppylinux-woof-CE/woof-CE/actions/runs/801826375, so we have GTK 2 (Slacko 7.x) and GTK 3 (Slacko 8.x) builds for bug hunting and comparison.

**Comment by peabee** [comment]
> Triggered https://github.com/puppylinux-woof-CE/woof-CE/actions/runs/801826375, so we have GTK 2 (Slacko 7.x) and GTK 3 (Slacko 8.x) builds for bug hunting and comparison.

Gtk3 aspects of Slacko64-8.1 seem much better although I feel quicksetup window is rather too wide....

@01micko - I notice that Slacko64-8.1 does not recognise Arch64 pkg.tar.zst packages....

Screenshot

**Comment by peabee** [comment]
p.s. why do the iso's and puppy.sfs of seemingly identical builds vary in size so much?

**Comment by dimkr** [comment]
> p.s. why do the iso's and puppy.sfs of seemingly identical builds vary in size so much?

You're probably talking about https://github.com/puppylinux-woof-CE/woof-CE/blob/testing/.github/workflows/build.yml#L79

**Comment by 01micko** [comment]
> I notice that Slacko64-8.1 does not recognise Arch64 pkg.tar.zst packages

@peabee are you using Slackware's `shared-mime-info`?

File: woof-code/rootfs-skeleton/usr/bin/pupdialog [hunk 1]
```diff
@@ -208,11 +208,27 @@ done
...
+	$FGCOL
+	font-size: 10pt;
+	font-style: normal;
+}
+box {
+	min-width: 600px;
+}
+button.radio:checked {
+	border: 3px solid ${FOREGROUND};
+}
```
Comment (01micko) [inline woof-code/rootfs-skeleton/usr/bin/pupdialog:229]:
```
pupdialog --backtitle "backtitle here" --background yellow --foreground grey20  --default-item tag2  --menu "this is blurb for menu dlg" 0 0 0 tag1 item1 tag2 item2 tag3 item3
```

pupdialog

**Comment by 01micko** [comment]
next stop .. `pmusic`

It might take a while.

**Comment by peabee** [comment]
> @peabee are you using Slackware's `shared-mime-info`?

No - my builds use common32 / 64 - version 1.10

Slacko64-8.1 is using very old version 1.2:
shared-mime-info-1.2-x86_64_s700|shared-mime-info|1.2-x86_64_s700||BuildingBlock|4976||shared-mime-info-1.2-x86_64_s700.pet||mimetype information|slackware64|14.2||
shared-mime-info_DEV-1.2-x86_64_s700|shared-mime-info_DEV|1.2-x86_64_s700||BuildingBlock|32||shared-mime-info_DEV-1.2-x86_64_s700.pet|+shared-mime-info|shared-mime-info development||||

**Comment by zigbert** [comment]
> next stop .. `pmusic`
> 
> It might take a while.

Looking forward to it :)

**Comment by 01micko** [comment]
> > @peabee are you using Slackware's `shared-mime-info`?
> 
> No - my builds use common32 / 64 - version 1.10
> 
> Slacko64-8.1 is using very old version 1.2:
> shared-mime-info-1.2-x86_64_s700|shared-mime-info|1.2-x86_64_s700||BuildingBlock|4976||shared-mime-info-1.2-x86_64_s700.pet||mimetype information|slackware64|14.2||
> shared-mime-info_DEV-1.2-x86_64_s700|shared-mime-info_DEV|1.2-x86_64_s700||BuildingBlock|32||shared-mime-info_DEV-1.2-x86_64_s700.pet|+shared-mime-info|shared-mime-info development||||

Thanks Peter, I'll correct that shortly.

**Comment by 01micko** [comment]
@peabee `shared-mime-info-1.10` is still too old. I had success with slackware's 2.1 but there is still work to do.

As this is off topic I created issue #2251

**Comment by 01micko** [comment]
@zigbert ziggy stardust is back

@dimkr 

widgets1

widgets2

widgets3

**Comment by zigbert** [comment]
Mmmm - you really get the hang of it ;) - Some stated that Stardust was 'impossible' in GTK3

I try to crack some bugs in my small drops of free time...
We are opening a new gallery in a near future. Not the same artwork that you are working on, but still...

**Comment by 01micko** [comment]
The problem is the png files as in gtk3 they don't stretch so you have to 'draw' images in pure CSS. The `stardust_bright` gtk3 theme is certainly _not_ a faithful reproduction of the gtk2 version but it's getting closer.
Actually, I opened this browser before I fixed the theme and you can see what I mean with the png files in the scrollbar.

scroll-bug


I'm going to commit it today and hopefully the 'touch' version as well then we go for the `stardust_dark`.

**Comment by 01micko** [comment]
@zigbert 

> We are opening a new gallery in a near future. Not the same artwork that you are working on, but still...

That's great news! Good Luck :four_leaf_clover:

**Comment by 01micko** [comment]
@dimkr - just thought I'd mention that commit I made to gtkdialog - [3c18c9e](https://github.com/puppylinux-woof-CE/gtkdialog/commit/3c18c9e0f1d00a30531fef16058b73f3a4fa49cf)

The maximum chars can easily be overidden with `width-chars`:
```xml
 <text width-chars="120">
```

**Comment by 01micko** [comment]
@dimkr - just thought I'd mention that commit I made to gtkdialog - [3c18c9e](https://github.com/puppylinux-woof-CE/gtkdialog/commit/3c18c9e0f1d00a30531fef16058b73f3a4fa49cf)

The maximimum chars can easily be overidden with `width-chars`:
```xml
 <text width-chars="120">
```

**Comment by 01micko** [comment]
@zigbert 

Since there are complexities in the theming of PMusic it will take me some time to get up to speed with gtk3. Probably several months. 

Also, pequalizer is giving problems - mostly due to the addition of pulseaudio. No big deal - it had a good run. Maybe I can make one based on pulse - but that's for another time. The upshot of that is that I won't be including pequalizer in future versions if pulseaudio is installed. 

PMusic runs for me just fine if I disable pequalizer and drop in a gtkdialog based on gtk2 - no problem if I build it without VTE support. I just renamed the exec to `gtk2dialog` so as not to conflict.

My request to you is to add this small patch to PMusic (and friends) so that I can use the alternate exec. The patch will be temporary until we get gtk3 theming in PMusic stable.

_PATCH REMOVED_ - see post below

TIA

**EDIT**
On further inspection I see `gtkdialog` is hard coded in more places than the patch mentions, especially in `ps` calls. Do they work with the variable `$GTKDIALOG`?

**Comment by 01micko** [comment]
@zigbert 

Here is a revised patch, seems to work ok.. no code is touched just hard coding replaced with variable - and `GTK3` test added in another spot.

```diff
diff -rupN a/usr/local/pmusic/box_input b/usr/local/pmusic/box_input
--- a/usr/local/pmusic/box_input	2020-07-09 08:32:28.000000000 +1000
+++ b/usr/local/pmusic/box_input	2021-05-09 11:37:10.763115364 +1000
@@ -2,7 +2,13 @@
 $APPDIR/func -reset_gtk
 export TEXTDOMAIN=pmusic
 export OUTPUT_CHARSET=UTF-8
-[ ! "$GTKDIALOG" ] && GTKDIALOG="gtkdialog" #puppy specific
+if [ ! "$GTKDIALOG" ];then #puppy specific
+	if [ "$GTKDIALOG_BUILD" = 'GTK3' ];then
+		export GTKDIALOG=gtk2dialog
+	else
+		export GTKDIALOG=gtkdialog
+	fi
+fi
 if [ -z "$FRAME" ]; then FRAME=" "; fi
 export WORKDIR=$HOME/.pmusic/tmp
 
diff -rupN a/usr/local/pmusic/func b/usr/local/pmusic/func
--- a/usr/local/pmusic/func	2021-03-03 12:38:35.000000000 +1000
+++ b/usr/local/pmusic/func	2021-05-09 11:24:33.253129431 +1000
@@ -141,7 +141,7 @@ pMusic will work just fine even if missi
 	   </tree>
 	  </vbox>
 	 </vbox>
-	</window>' | gtkdialog -s &
+	</window>' | $GTKDIALOG -s &
 	;;
 
 -visualization)
@@ -773,7 +773,7 @@ $(gettext 'Requested filter:') $TMP"
 	#first quit guis
 	TMP="$(ps -eo pid,command)"
 	PS=$(grep -iE "pmusic|aplay|ffplay" <<< "$TMP")
-	for I in `grep -F gtkdialog <<< "$PS" | grep -vF radiohits | awk '{print $1}' 2>> $STDERR`; do kill -9 $I 2>> $STDERR; done #do not kill radiohits grabber 
+	for I in `grep -F $GTKDIALOG <<< "$PS" | grep -vF radiohits | awk '{print $1}' 2>> $STDERR`; do kill -9 $I 2>> $STDERR; done #do not kill radiohits grabber 
 	#stop audio
 	pmusic -s stop
 	rm -f $WORKDIR/gui_is_running #ensure to flag 'no gui'
@@ -797,7 +797,7 @@ $(gettext 'Requested filter:') $TMP"
 	for I in `grep -F index <<< "$PS" | awk '{print $1}' 2>> $STDERR`; do kill -9 $I 2>> $STDERR; done
 	for I in `grep -F aplay <<< "$PS" | awk '{print $1}' 2>> $STDERR`; do kill -9 $I 2>> $STDERR; done
 	for I in `grep -F playing <<< "$PS" | awk '{print $1}' 2>> $STDERR`; do kill -9 $I 2>> $STDERR; done
-	for I in `grep -F gtkdialog <<< "$PS" | grep -vF radiohits | awk '{print $1}' 2>> $STDERR`; do kill -9 $I 2>> $STDERR; done
+	for I in `grep -F $GTKDIALOG <<< "$PS" | grep -vF radiohits | awk '{print $1}' 2>> $STDERR`; do kill -9 $I 2>> $STDERR; done
 	for I in `grep -vF radiohits <<< "$PS" | awk '{print $1}' 2>> $STDERR`; do kill -9 $I 2>> $STDERR; done
 	killall cdda2wav 2>> $STDERR
 	#reset info about search filters. Next run will use settings from preferences
diff -rupN a/usr/local/pmusic/gui_trackinfo b/usr/local/pmusic/gui_trackinfo
--- a/usr/local/pmusic/gui_trackinfo	2020-07-09 08:32:28.000000000 +1000
+++ b/usr/local/pmusic/gui_trackinfo	2021-05-09 11:30:14.783123061 +1000
@@ -90,7 +90,7 @@ S='
   <button tooltip-text="'$(gettext 'Search for alternative meta information based on ARTIST and TITLE. The search-string does not require complete Artist/Title information, but you MUST use complete words.')'">
    <input file stock="gtk-find"></input>
    <action>'$APPDIR'/func_trackinfo "'${MYPID}'" -search_alternatives</action>
-   <action>gtkdialog -f '$WORKDIR'/xml_pmusic_trackinfo'${MYPID}'_alternativ_id3 &</action>
+   <action>'$GTKDIALOG' -f '$WORKDIR'/xml_pmusic_trackinfo'${MYPID}'_alternativ_id3 &</action>
   </button>
  </hbox>
  
diff -rupN a/usr/local/pmusic/pmusic b/usr/local/pmusic/pmusic
--- a/usr/local/pmusic/pmusic	2021-03-03 12:38:35.000000000 +1000
+++ b/usr/local/pmusic/pmusic	2021-05-09 12:04:41.803084723 +1000
@@ -23,7 +23,11 @@ export OUTPUT_CHARSET=UTF-8
 #---
 export AUDIO_FORMATS="\.3gp$|\.aac$|\.ac3$|\.act$|\.aif$|\.aiff$|\.amr$|\.ape$|\.asf$|\.au$|\.dss$|\.flac$|\.gsm$|\.kar$|\.m4a$|\.m4b$|\.mid$|\.mmf$|\.mp1$|\.mp2$|\.mp3$|\.mp4$|\.mpc$|\.oga$|\.ogg$|\.opus$|\.ra$|\.raw$|\.rm$|\.shn$|\.sln$|\.tta$|\.voc$|\.wav$|\.webm$|\.wma$|\.wv$"
 type pequalizer > /dev/null 2>&1 && pequalizer -s & #setup initial presets
-export GTKDIALOG=gtkdialog
+if [ "$GTKDIALOG_BUILD" = 'GTK3' ];then
+	export GTKDIALOG=gtk2dialog
+else
+	export GTKDIALOG=gtkdialog
+fi
 export TXT_SEARCH="$(gettext 'Searching more info')"
 export TXT_NO_WWW="$(gettext 'Not able to connect to internet')"
 export SF='space-expand="false" space-fill="false"'
@@ -296,8 +300,8 @@ while [ $# != 0 ]; do
 				if type cp > /dev/null 2>&1 && type awk > /dev/null 2>&1 && type sed > /dev/null 2>&1 && type diff > /dev/null 2>&1; then echo -e "|OK|" >> $WORKDIR/dependencies; else echo -e "|MISSING|" >> $WORKDIR/dependencies; fi
 				#gtkdialog
 				TMP=`$GTKDIALOG --version 2>> $STDERR | grep -F version | cut -d' ' -f3`
-				echo -n '    gtkdialog                       |http://www.murga-linux.com/puppy/viewtopic.php?t=38608|| >= 0.8.4' >> $WORKDIR/dependencies
-				if type gtkdialog > /dev/null 2>&1 && [[ $((10#${TMP//./})) -ge 84 ]]; then echo -e "|OK|" >> $WORKDIR/dependencies; else echo -e "|MISSING|" >> $WORKDIR/dependencies; fi
+				echo -n "    $GTKDIALOG                       |http://www.murga-linux.com/puppy/viewtopic.php?t=38608|| >= 0.8.4" >> $WORKDIR/dependencies
+				if type $GTKDIALOG > /dev/null 2>&1 && [[ $((10#${TMP//./})) -ge 84 ]]; then echo -e "|OK|" >> $WORKDIR/dependencies; else echo -e "|MISSING|" >> $WORKDIR/dependencies; fi
 				#ffmpeg is not that straight forward since compiling from git gives another output
 				TMP=`ffmpeg -version 2>> $STDERR | grep -iF 'ffmpeg' | cut -d' ' -f3 | cut -d'-' -f1`
 				echo -n '    ffmpeg                          |https://ffmpeg.org/|| >= 1.2  ' >> $WORKDIR/dependencies
@@ -423,7 +427,7 @@ if [ -s $WORKDIR/MY_PID ]; then
 fi
 echo $$ > $WORKDIR/MY_PID #store for next instance (new PMUSIC_FRONTEND or new startup without quit)
 PS="$(LC_ALL=C ps -eo pid,s,command)"
-for I in `LC_ALL=C grep -i pmusic <<< "$PS" | grep -F gtkdialog | grep -vF radiohits | awk '{print $1}'`; do kill -9 $I 2>> $STDERR; done
+for I in `LC_ALL=C grep -i pmusic <<< "$PS" | grep -F $GTKDIALOG | grep -vF radiohits | awk '{print $1}'`; do kill -9 $I 2>> $STDERR; done
 
 if [ "$NOGUI" != "true" ]; then
 	#check for data storage

```

~The very last change is the only one that's concerning because gtkdialog may refer to itself internally as 'gtkdialog', no matter the name of the exec.~

EDIT: Yes that strike thru was a problem, revised patch posted.

**Comment by 01micko** [comment]
Ok, patch works - built from woof, tested in qemu - great sound!

qemu-pmusic(1)

**Comment by dimkr** [comment]
~~Where is gtk2dialog coming from? Maybe we should change the gtkdialog petbuild so it builds against both GTK+ 2 and 3.~~

#2254

**Comment by zigbert** [comment]
@01micko Of course add your patch. 
I first misunderstood, and thought you wanted me to add the changes (for some reason...) - Therefor my initial commit this morning.

**Comment by 01micko** [comment]
@zigbert no problem - I see you added part of it - I'll complete the job :+1:

**Comment by 01micko** [comment]
[slacko64-8.2](https://forum.puppylinux.com/viewtopic.php?p=24908#p24908) built on bare metal from the latest `gtk3` branch available for dowmload.

File: woof-code/rootfs-packages/ptheme/usr/sbin/ptheme [hunk 2]
```diff
@@ -46,7 +46,7 @@ build_menus (){
 	MENUITEMS_JWM_COLOR="$(find /usr/share/jwm/themes/ -mindepth 1 -maxdepth 1 -iname "*jwmrc" -printf "<menuitem label=\"%f\"><action>echo %f > $WORKDIR/ptheme-jwm</action><action>refresh:PTHEME_JWM_COLOR</action></menuitem>" | sort | sed -e 's/-jwmrc//g' &)"
 	MENUITEMS_JWM_TRAY="$(build_menu /usr/share/jwm/tray_templates $WORKDIR/ptheme-jwm_tray PTHEME_JWM_TRAY)"
 	MENUITEMS_JWM_BUTTONS="$(build_menu /usr/share/jwm/themes_window_buttons/ $WORKDIR/ptheme-jwm_buttons PTHEME_JWM_BUTTONS)"
-	MENUITEMS_MOUSE="$(find -L $HOME/.icons -mindepth 1 -maxdepth 1 -type d -printf "<menuitem label=\"%f\"><action>echo %f > $WORKDIR/ptheme-mouse</action><action>refresh:PTHEME_MOUSE</action></menuitem>" | sort | grep -vE "ROX|default" &)"
+	MENUITEMS_MOUSE="$(find $HOME/.icons -mindepth 1 -maxdepth 1 -type d -printf "<menuitem label=\"%f\"><action>echo %f > $WORKDIR/ptheme-mouse</action><action>refresh:PTHEME_MOUSE</action></menuitem>" | sort | grep -vE "ROX|default" &)"
```
Comment (dimkr) [inline woof-code/rootfs-packages/ptheme/usr/sbin/ptheme:49]: Why is this -L removed?

Comment (dimkr) [reply woof-code/rootfs-packages/ptheme/usr/sbin/ptheme:49]: https://github.com/puppylinux-woof-CE/woof-CE/commit/0e9de2dcbac3daf4aa53164aa782c0a72f7d1d2f

Comment (01micko) [reply woof-code/rootfs-packages/ptheme/usr/sbin/ptheme:49]: PEBKAC (here) - fixed now
58435b3

**Comment by 01micko** [comment]
We're getting closer  -

gtk3-apps

**Comment by dimkr** [comment]
> We're getting closer -

Nice!

I'm seeing some issues with Flat-grey-rounded now - some widgets are too small with straight corners now, and stuff like MATE panel doesn't work well (window list item color doesn't change on hover).

**Comment by 01micko** [comment]
> > We're getting closer -
> 
> Nice!
> 
> I'm seeing some issues with Flat-grey-rounded now - some widgets are too small with straight corners now, and stuff like MATE panel doesn't work well (window list item color doesn't change on hover).

There are  _lots_  of issues! I haven't committed anything yet because there are tiny issues in lots of different apps. I'm still learning how to control CSS with gtk. It's similar but definitely not as forgiving as web development.

**Comment by 01micko** [comment]
BTW @dimkr what distro are you running mate on?

**Comment by dimkr** [comment]
> BTW @dimkr what distro are you running mate on?

Ubuntu MATE. But it doesn't matter - even stuff like the "open file" dialog is messed and it shouldn't be an Ubuntu-specific problem. You can see that the padding is gone, and the square widgets look out of place.

(Maybe, every "flat" widget is messed up because the CSS selector has changed)

flat

**Comment by 01micko** [comment]
I'll get something going with MATE and take a look.

**Comment by dimkr** [comment]
@01micko Why not merge it now?

**Comment by 01micko** [comment]
Be my guest @dimkr

**Review by dimkr — APPROVED** [review]

**Pull request merged**